packages/scm/src/browser/decorations/scm-decorations-service.ts

```diff
@@ -15,64 +15,82 @@
+        @inject(DirtyDiffDecorator) protected readonly decorator: DirtyDiffDecorator,
         @inject(ScmService) protected readonly scmService: ScmService,
         @inject(EditorManager) protected readonly editorManager: EditorManager,
-        @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider) {
-        this.diffComputer = new DiffComputer();
-        this.editorManager.onCreated(async editor => this.applyEditorDecorations(editor.editor));
-        this.scmService.onDidAddRepository(repository => repository.provider.onDidChange(() => {
-            const editor = this.editorManager.currentEditor;
-            if (editor) {
-                if (this.dirtyState) {
-                    this.applyEditorDecorations(editor.editor);
-                    this.dirtyState = false;
-                } else {
-                    /** onDidChange event might be called several times one after another, so need to prevent repeated events. */
-                    setTimeout(() => {
-                        this.dirtyState = true;
-                    }, 500);
-                }
+        @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider
+    ) {
+        const updateTasks = new Map<EditorWidget, () => void>();
+        this.editorManager.onCreated(editorWidget => {
+            const { editor } = editorWidget;
+            if (editor.uri.scheme !== 'file') {
+                return;
             }
-        }));
-        this.scmService.onDidChangeSelectedRepository(() => {
-            const editor = this.editorManager.currentEditor;
-            if (editor) {
-                this.applyEditorDecorations(editor.editor);
+            const toDispose = new DisposableCollection();
+            const updateTask = this.createUpdateTask(editor);
+            updateTasks.set(editorWidget, updateTask);
+            toDispose.push(editor.onDocumentContentChanged(() => updateTask()));
+            editorWidget.disposed.connect(() => {
+                updateTasks.delete(editorWidget);
+                toDispose.dispose();
+            });
+            updateTask();
+        });
+        const runUpdateTasks = () => {
+            for (const updateTask of updateTasks.values()) {
+                updateTask();
             }
+        };
+        this.scmService.onDidAddRepository(({ provider }) => {
+            provider.onDidChange(runUpdateTasks);
+            provider.onDidChangeResources?.(runUpdateTasks);
         });
+        this.scmService.onDidChangeSelectedRepository(runUpdateTasks);
     }
 
     async applyEditorDecorations(editor: TextEditor): Promise<void> {
         const currentRepo = this.scmService.selectedRepository;
         if (currentRepo) {
             try {
-                const uri = editor.uri.withScheme(currentRepo.provider.id).withQuery(`{"ref":"", "path":"${editor.uri.path.toString()}"}`);
+                const query = { path: editor.uri['codeUri'].fsPath, ref: '~' };
```

Comment:
I'll try to explain this change in detail.

Let's start with the original line of code:

https://github.com/eclipse-theia/theia/blob/df6341a4aecec7846cf72cf9f1fdd1c2b5bc5e36/packages/scm/src/browser/decorations/scm-decorations-service.ts#L62

As far as I can tell, the format of the query in this URI is specific to `vscode.git`, and is defined by functions `fromGitUri` and `toGitUri` in [extensions/git/src/uri.ts](https://github.com/microsoft/vscode/blob/9f4bcdde684bc54e311ea49cb469771796094e7d/extensions/git/src/uri.ts). Therefore, it does not seem to support SCM providers other than `vscode.git`. I have actually installed an SVN SCM provider and can confirm that, on the current master, there are no SCM decorations shown in the editor gutter for SVN.

The patch keeps this baseline behavior, but fixes a couple of issues with the URI to ensure that it actually works in all cases with `vscode.git`:

1. As can be seen in the implementation of `toGitUri` in [extensions/git/src/uri.ts](https://github.com/microsoft/vscode/blob/9f4bcdde684bc54e311ea49cb469771796094e7d/extensions/git/src/uri.ts), `vscode.git` uses `fsPath` rather than `path` in the query.

2. As `fsPath` may contain backslashes on Windows, they need to be escaped with extra `\` in JSON. Therefore, rather than trying to construct the JSON string manually, `JSON.stringify` is now used to handle the escaping automatically, just as `vscode.git` does in its `toGitUri` function. This fixes the issue on Windows that you previously discovered.

3. `'~'` is now used as the ref instead of the empty string to ensure that it is correctly mapped to either `''` or `'HEAD'` by the `sanitizeRef` function in [extensions/git/src/fileSystemProvider.ts](https://github.com/microsoft/vscode/blob/9f4bcdde684bc54e311ea49cb469771796094e7d/extensions/git/src/fileSystemProvider.ts).

> There is code to get the file system path in file-uri.ts.

Yes, but:

1. `file-uri.ts` was node-only at the time this PR was created.

2. There is some post-processing of `fsPath` in `file-uri.ts` in the Windows case, while `vscode.git` expects the unmodified `fsPath`.

Therefore, I believe that the use of `uri['codeUri'].fsPath` is correct in this case.

Reply:
I've looked at the corresponding code in VS Code and they seem to be using the "QuickDIffProvider" API for getting the proper URIs. There is no implementation of that concept in our Scm module as far as I can see. So this is a hack to make the feature work for git. As this is preexisting code, that's fair enough, but could you please add a comment in the code? There's just no way anyone figures this out on their own if the read this in isolation.

Reply:
As requested, I've added the comment in https://github.com/eclipse-theia/theia/pull/13104/commits/c98d16098d26f487399226ff727efb82ad07026b.